tmt/steps/report/reportportal.py

```diff
@@ -91,77 +139,222 @@ class ReportReportPortal(tmt.steps.report.ReportPlugin[ReportReportPortalData]):
 
-        assert self.step.plan.name is not None
-        launch_name = self.data.launch or self.step.plan.name
-
-        url = f"{server}/api/{self.DEFAULT_API_VERSION}/{project}"
+        url = f"{endpoint}/api/{self.DEFAULT_API_VERSION}/{project}"
         headers = {
             "Authorization": "bearer " + token,
             "accept": "*/*",
             "Content-Type": "application/json"}
 
-        envar_pattern = self.data.exclude_variables
+        launch_time = str(int(time() * 1000))
+
+        # to support idle tests
+        executed = False
+        if len(self.step.plan.execute.results()) > 0:
+            launch_time = self.step.plan.execute.results()[0].start_time
+            executed = True
+
+        # Create launch, suites (if "--suite_per_plan") and tests;
+        # or report to existing launch/suite if its id is given
+        launch_uuid = self.get("launch_uuid") or self.step.plan.my_run.rp_uuid
+        suite_uuid = self.get("suite_uuid")
+
+        launch_id = self.get("upload_to_launch")
+        suite_id = self.get("upload_to_suite")
+
+        suite_per_plan = self.get("suite_per_plan")
+        launch_per_plan = self.get("launch_per_plan")
+        if not launch_per_plan and not suite_per_plan:
+            launch_per_plan = True      # by default
+        elif launch_per_plan and suite_per_plan:
+            raise tmt.utils.ReportError(
+                "The options '--launch-per-plan' and "
+                "'--suite-per-plan' are mutually exclusive. Choose one of them only.")
+
+        create_launch = not (launch_uuid or launch_id or suite_uuid or suite_id)
+        create_suite = suite_per_plan and not (suite_uuid or suite_id)
+
+        launch_url = ""
+        launch_name = self.get("launch") or self.step.plan.name
+
+        launch_rerun = self.get("launch_rerun")
+        envar_pattern = self.get("exclude-variables") or "$^"
+        defect_type = self.get("defect_type")
+
         attributes = [
             {'key': key, 'value': value[0]}
             for key, value in self.step.plan._fmf_context.items()]
-        launch_time = self.step.plan.execute.results()[0].start_time
+
+        if suite_per_plan:
+            # Get common attributes from all plans
+            merged_plans = [{key: value[0] for key, value in plan._fmf_context.items()}
+                            for plan in self.step.plan.my_run.plans]
+            result_dict = merged_plans[0]
+            for current_plan in merged_plans[1:]:
+                tmp_dict = {}
+                for key, value in current_plan.items():
+                    if key in result_dict and result_dict[key] == value:
+                        tmp_dict[key] = value
+                result_dict = tmp_dict
+            launch_attributes = [
+                {'key': key, 'value': value}
+                for key, value in result_dict.items()]
+        else:
+            launch_attributes = attributes.copy()
```

Comment:
Something to consider, maybe: I'm worried that the resulting `go()` method will be long and hard to follow and maintain. Would it make sense to split it (or most of it) into smaller functions, working together, into helper functions designed for one particular step of the bigger picture?

E.g. the `if suite_per_plan` above. It might be turned into a nifty function, something like the following:

```python
def construct_launch_attributes(self, suite_per_plan: bool, attributes: Dict[str, str]) -> Dict[str, str]:
    if not suite_per_plan:
        return attributes.copy()

    merged_plans = [{key: value[0] for key, value in plan._fmf_context.items()}
                    for plan in self.step.plan.my_run.plans]
    result_dict = merged_plans[0]
    ...
```

In this method, only `launch_atrributes = self.construct_launch_attributes(suite_per_plan, attributes)` would remain. `go()` would be easier to follow, dealing with logical steps on a higher level, `construct_launch_attributes` would be much easier to test with unit tests as it would define its inputs and return value - you cannot reasonably test the creation of launch attributes now -, and `go()` wouldn't be full of various one-shot variables.

And there are multiple such blocks that could be isolated, e.g. `dt_locator` - take the code under `if detect_type`, make it fetch and return the defect type, and `dt_locator = self.get_defect_type(...) if detect_type else None`.

I don't know how far in your work you are, but I'm worried we're getting a bit too close to a second `DiscoverFmf.go()`.

@psss what do you think?

Reply:
> Something to consider, maybe: I'm worried that the resulting `go()` method will be long and hard to follow and maintain. Would it make sense to split it (or most of it) into smaller functions, working together, into helper functions designed for one particular step of the bigger picture?

Yes, agreed, this would definitely help with readability and maintainability as well, +1 from me.

Reply:
@4N0body5 is this resolved?

Reply:
It's not, and it's probably too late now to start refactoring...